### PR TITLE
Issue 112 - Refactor frogger to new style, model simulation

### LINKED ISSUE #112: Refactor model_simulation code

### PR DESCRIPTION
**Pull request by M-Kusumgar**
Just does the model simulation file. I have taken a couple of liberties. I will point out the points of interest with comments in the diff. Resolves #112

### CODE REVIEW

File: inst/include/model_simulation.hpp [hunk 1]
```diff
@@ -236,96 +210,76 @@ void run_art_progression_and_mortality(int hiv_step,
...
+	      }
       }
     }
 
     // calculate number on ART at end of ts, based on number or percent
-    if (pars.base.options.proj_period_int == internal::PROJPERIOD_MIDYEAR &&
-	dt * (hiv_step + 1) < 0.5) {
+    real_type art_interp_w = p_op.dt * (hiv_step + 1.0);
+    if (p_op.proj_period_int == internal::PROJPERIOD_MIDYEAR && art_interp_w < 0.5) {
+      // case: both values are numbers
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:259]: point of interest: i have split up the `if else if`s into their own if statements it breaks them up and gives a cleaner code structure in my opinion

File: inst/include/model_simulation.hpp [hunk 2]
```diff
@@ -236,96 +210,76 @@ void run_art_progression_and_mortality(int hiv_step,
...
+          i_ba.Xart_15plus += n_ba.h_art_adult(hu, hm, ha, g) +
+                              p_op.dt * i_ba.gradART(hu, hm, ha, g);
+	      }
       }
     }
 
     // calculate number on ART at end of ts, based on number or percent
-    if (pars.base.options.proj_period_int == internal::PROJPERIOD_MIDYEAR &&
-	dt * (hiv_step + 1) < 0.5) {
+    real_type art_interp_w = p_op.dt * (hiv_step + 1.0);
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:257]: point of interest: hopefully this is okay to define outside of any of the if statements, this was previously in the else branch

File: inst/include/model_simulation.hpp [hunk 3]
```diff
@@ -334,96 +288,69 @@ void run_h_art_initiation(int hiv_step,
...
-	      intermediate.base.artinit_hahm;
+    for (int ha = p_op.hIDX_15PLUS; ha < ss_b.hAG; ++ha) {
+      for (int hm = i_ba.anyelig_idx; hm < ss_b.hDS; ++hm) {
+        if (i_ba.artelig_hm(hm) > 0.0) {
+          i_ba.artinit_hahm = i_ba.artinit_hm(hm) *
+	                            i_ba.artelig_hahm(hm, ha - p_op.hIDX_15PLUS) /
+	                            i_ba.artelig_hm(hm);
+          i_ba.artinit_hahm = std::max(i_ba.artinit_hahm, i_ba.artelig_hahm(hm, ha - p_op.hIDX_15PLUS));
+          i_ba.artinit_hahm = std::max(i_ba.artinit_hahm,
+                                       n_ba.h_hiv_adult(hm, ha, g) + p_op.dt * i_ba.grad(hm, ha, g));
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:350]: point of interest: have replaced the ifs with `std::max`s it achieves the same thing

File: inst/include/model_simulation.hpp [hunk 4]
```diff
@@ -511,71 +444,61 @@ void run_remove_p_hiv_deaths(int hiv_step,
...
+  const auto& p_tx = pars.base.art;
+  const auto& p_op = pars.base.options;
+  auto& i_ba = intermediate.base;
+
+  i_ba.everARTelig_idx = p_tx.idx_hm_elig(t) < ss.hDS ? p_tx.idx_hm_elig(t) : ss.hDS;
+  i_ba.anyelig_idx = p_tx.idx_hm_elig(t);
+
+  internal::run_calculate_incidence_rate<ModelVariant>(
+    t, pars, state_curr, state_next, intermediate
+  );
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:462]: point of interest: new function arg style, yay or nay?

File: inst/include/model_simulation.hpp [hunk 5]
```diff
@@ -334,96 +288,69 @@ void run_h_art_initiation(int hiv_step,
...
-	      intermediate.base.artinit_hahm;
+    for (int ha = p_op.hIDX_15PLUS; ha < ss_b.hAG; ++ha) {
+      for (int hm = i_ba.anyelig_idx; hm < ss_b.hDS; ++hm) {
+        if (i_ba.artelig_hm(hm) > 0.0) {
+          i_ba.artinit_hahm = i_ba.artinit_hm(hm) *
+	                            i_ba.artelig_hahm(hm, ha - p_op.hIDX_15PLUS) /
+	                            i_ba.artelig_hm(hm);
+          i_ba.artinit_hahm = std::min(i_ba.artinit_hahm, i_ba.artelig_hahm(hm, ha - p_op.hIDX_15PLUS));
+          i_ba.artinit_hahm = std::min(i_ba.artinit_hahm,
+                                       n_ba.h_hiv_adult(hm, ha, g) + p_op.dt * i_ba.grad(hm, ha, g));
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:350]: point of interest: does same thing as the if statements

File: inst/include/model_simulation.hpp [hunk 6]
```diff
@@ -334,96 +288,69 @@ void run_h_art_initiation(int hiv_step,
...
+        i_ba.artnum_hts = i_ba.artcov_hts * (i_ba.Xart_15plus + i_ba.Xartelig_15plus);
       }
     }
 
     // Desired number to initiate on ART
-    intermediate.base.artinit_hts =
-        intermediate.base.artnum_hts > intermediate.base.Xart_15plus ?
-        intermediate.base.artnum_hts -
-        intermediate.base.Xart_15plus : 0.0;
+    i_ba.artinit_hts = std::max(i_ba.artnum_hts - i_ba.Xart_15plus, 0.0);
```
Comment (M-Kusumgar) [inline inst/include/model_simulation.hpp:322]: point of interest: max instead of if statement

**Review by mwalte10 — APPROVED** [review]
Looks good!

File: inst/include/model_simulation.hpp (hunk 1 above)
Comment (jeffeaton) [reply inst/include/model_simulation.hpp:259]: Per discussion — revert this back to `else if`

**Review by jeffeaton — CHANGES_REQUESTED** [review]

**Pull request merged**